src/loader.ts

```diff
@@ -18,29 +18,136 @@
  * @packageDocumentation
  */
 
+import * as path from 'path';
+import * as semver from 'semver';
+
 /**
  * Import function signature type's definition.
  */
 import {HandlerFunction} from './functions';
 
+// Dynamic import function required to load user code packaged as an
+// ES module is only available on Node.js v13.2.0 and up.
+//   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#browser_compatibility
+// Exported for testing.
+export const MIN_NODE_VERSION_ESMODULES = '13.2.0';
+
+/**
+ * Determines format of the given module (CommonJS vs ES module).
+ *
+ * Implements "algorithm" described at:
+ *   https://nodejs.org/api/packages.html#packages_type
+ *
+ * In words:
+ *   1. A module with .mjs extension is an ES module.
+ *   2. A module with .clj extension is CommonJS.
+ *   3. A module with .js extensions where...
+ *     a. Nearest package.json's with "type": "module" is an ES
+ *        module.
+ *     b. Otherwise, it is CommonJS.
+ *
+ * @returns {string} Module format ('commonjs' or 'module')
+ */
+function moduleFormat(modulePath: string): 'commonjs' | 'module' {
```

Comment:
newbie question: Do we prefer to use an external package when possible or is there any reason to limit external deps in this repo?

The list of deps for this package was already pretty slim, and I had small amount of hesitation adding new ones. The package you linked seems to be well tested, but I can imagine going the other route and beefing up the unit tests for this function rather than adding a dep.

Reply:
If an external package seems useful, not abandoned, and genetic for the problem, then use it. I like to remove code that isn't unique to this library / i.e. could be used in other packages as package dependencies. We do keep dependencies small in general.

Reply:
I took a closer look at https://www.npmjs.com/package/is-file-esm 's source code. I'm being very nitpicky, but it seems to  me that it has bit more complex implementation than what I have here. To be fair, it does more, but we don't need those extra functiond ality.

I lean towards keeping couple of lines and keep this logic to ourselves, but that's just me coming from go-world where I'm used to doing things like this. If you think we should opt-in to use that package, I'll be happy to revise.

Reply:
Whoops. Discovered that `is-file-esm` makes uses of [`queueMicroStack`](https://nodejs.org/api/globals.html#globals_queuemicrotask_callback) in https://github.com/davidmarkclements/is-file-esm/blob/master/index.js#L9 which is only available on NodeJS 11 and up.

I think functions framework needs to work for nodejs10 for the foreseeable future. Does this make using the library a deal breaker?

Reply:
Oh, hmmm.

Let's keep the lines here then. 

---

Aside, Node 12 should be latest LTS but we need to update min engine to that. But I'm not sure if that will take a bit.